explorer/interpreter/interpreter.cpp

```diff
@@ -706,8 +706,11 @@ auto Interpreter::ConvertStructToClass(
   }
   auto* converted_init_struct =
       arena_->New<StructValue>(std::move(struct_values));
+  Nonnull<const NominalClassValue** const> class_value_ptr =
+      base_instance ? (*base_instance)->class_value_ptr()
+                    : arena_->New<const NominalClassValue*>();
   return arena_->New<NominalClassValue>(inst_class, converted_init_struct,
-                                        base_instance);
+                                        base_instance, class_value_ptr);
```

Comment:
Now that I'm understanding what's going on a little better, how about updating the vptr here:

```
auto new_child = arena_->New<NominalClassValue>(inst_class, converted_init_struct,
                                        base_instance, class_value_ptr);
*class_value_ptr = new_child;
return new_child;
```

This gets rid of the semantic where the constructor modifies the pointer, which feels like modifying unowned state and has been making it harder for me to understand the flow.

Reply:
We could, but classes are constructed in two locations: here and in `SetFieldImpl`. One looks like modifying un-owned data, the other creates the risk of forgetting/misusing that (non-obvious) logic where class values are constructed. I'm ok with either though.